Sources/SwiftFusion/Nonlinear/NonlinearFactorGraph.swift

```diff
@@ -0,0 +1,114 @@
+import TensorFlow
+
+public struct NonlinearFactorGraph {
+  var factor1s: [Factor1] = []
+  var factor2s: [Factor2] = []
+
+  var factorCount: Int { factor1s.count + factor2s.count }
+
+  public init() {}
+
+  public mutating func add(
+    _ value1Index: Int, _ error: @escaping @differentiable (Pose2) -> Pose2, _ noise: NoiseModel
+  ) {
+    factor1s.append(Factor1(value1Index: value1Index, error: error, noise: noise))
+  }
+
+  public mutating func add(
+    _ value1Index: Int, _ value2Index: Int,
+    _ error: @escaping @differentiable (Pose2, Pose2) -> Pose2,
+    _ noise: NoiseModel
+  ) {
+    factor2s.append(Factor2(
+      value1Index: value1Index, value2Index: value2Index, error: error, noise: noise))
+  }
+
+  public func loss(at point: [Pose2]) -> Double {
+    factor1s.map { $0.noise.loss($0.error(point[$0.value1Index])) }.reduce(0, +) +
+      factor2s.map { $0.noise.loss($0.error(point[$0.value1Index], point[$0.value2Index])) }.reduce(0, +)
+  }
+
+  /// Returns an nx1 matrix (a vector) whose entries are the factors' errors at `point`.
+  public func errors(at point: [Pose2]) -> Tensor<Double> {
+    var vector = Tensor<Double>(zeros: [3 * factorCount, 1])
+    var currentFactorIndex: Int = 0
+    for factor in factor1s {
+      let error = factor.error(point[factor.value1Index])
+      vector[3 * currentFactorIndex, 0] = Tensor(error.rot.theta)
```

Comment:
Not using any of the Vector -> Tensor conversion functions yet.

Reply:
I think we might never go there: just insist a factor returns a tensor?

Reply:
Oh yes, that's a good idea.